src/client/mod.rs

```diff
@@ -137,48 +189,71 @@ impl InfluxDbClient {
             let read_query = query.get();
-            let encoded: String = form_urlencoded::Serializer::new(String::new())
-                .append_pair("db", self.database_name())
-                .append_pair("q", &read_query)
-                .finish();
-            let http_query_string = format!(
-                "{url}/query?{encoded}",
-                url = self.database_url(),
-                encoded = encoded
-            );
+
+            let mut url = match Url::parse_with_params(
+                format!("{url}/query", url = self.database_url()).as_str(),
+                basic_parameters,
+            ) {
+                Ok(url) => url,
+                Err(err) => {
+                    let error = InfluxDbError::UrlConstructionError {
+                        error: format!("{}", err),
+                    };
+                    return Box::new(future::err::<String, InfluxDbError>(error));
+                }
+            };
+            url.query_pairs_mut().append_pair("q", &read_query.clone());
+
             if read_query.contains("SELECT") || read_query.contains("SHOW") {
-                Client::new().get(http_query_string.as_str())
+                Client::new().get(url)
             } else {
-                Client::new().post(http_query_string.as_str())
+                Client::new().post(url)
             }
         } else if let Some(write_query) = any_value.downcast_ref::<InfluxDbWriteQuery>() {
-            Client::new()
-                .post(
-                    format!(
-                        "{url}/write?db={db}{precision_str}",
-                        url = self.database_url(),
-                        db = self.database_name(),
-                        precision_str = write_query.get_precision_modifier()
-                    )
-                    .as_str(),
-                )
-                .body(query.get())
+            let mut url = match Url::parse_with_params(
+                format!("{url}/write", url = self.database_url()).as_str(),
+                basic_parameters,
+            ) {
+                Ok(url) => url,
+                Err(err) => {
+                    let error = InfluxDbError::InvalidQueryError {
+                        error: format!("{}", err),
+                    };
+                    return Box::new(future::err::<String, InfluxDbError>(error));
+                }
+            };
+            url.query_pairs_mut()
+                .append_pair("precision", &write_query.get_precision());
+            Client::new().post(url).body(query.get())
         } else {
             unreachable!()
         };
-
         Box::new(
             client
                 .send()
+                .map_err(|err| InfluxDbError::ConnectionError { error: err })
+                .and_then(
+                    |res| -> future::FutureResult<reqwest::r#async::Response, InfluxDbError> {
+                        match res.status() {
+                            StatusCode::UNAUTHORIZED => {
```

Comment:
These cases can be combined as `StatusCode::UNAUTHORIZED | StatusCode::FORBIDDEN => futures::future::err(InfluxDbError::AuthorizationError)`. I think the brackets are not necessary :)

Reply:
But they are distinct. One covers wrong auth or no credentials and the other one covers missing privileges for a correct user.
I will remove the brackets.

Reply:
Oops, sorry, I thought they had the same body.

Reply:
cargo fmt wants those brackets there. Removing them would fail the linter.

Reply:
That's strange. let's keep them then :)